pandas/_libs/index.pyx

```diff
@@ -270,7 +270,14 @@ cdef class IndexEngine:
             Py_ssize_t i, j, n, n_t, n_alloc
 
         self._ensure_mapping_populated()
-        values = np.array(self._get_index_values(), copy=False)
+        if targets[isnaobj(targets)].size > 0:
+            new_targets = [0 if checknull(t) else t for t in targets]
```

Comment:
Can you not just reuse the mask used for branching? So before the branch something like `mask = isnaojb(targets)` and then here do `targets[mask] = 0` ?

Reply:
Ideally want to avoid the list comprehension

Reply:
Sounds good. I will also use a mask to avoid the list comprehension in filling the values array with 0s.